src/System.Text.Json/src/System/Text/Json/JsonUtf8Reader.cs

```diff
@@ -0,0 +1,1274 @@
+        /// <param name="isFinalBlock">True when the input span contains the entire data to process.
+        /// Set to false only if it is known that the input span contains partial data with more data to follow.</param>
+        /// <param name="state">If this is the first call to the ctor, pass in a default state. Otherwise,
+        /// capture the state from the previous instance of the <see cref="JsonUtf8Reader"/> and pass that back.</param>
+        /// <remarks>
+        /// Since this type is a ref struct, it is a stack-only type and all the limitations of ref structs apply to it.
+        /// This is the reason why the ctor accepts a <see cref="JsonReaderState"/>.
+        /// </remarks>
+        public JsonUtf8Reader(ReadOnlySpan<byte> jsonData, bool isFinalBlock, JsonReaderState state)
+        {
+            _buffer = jsonData;
+
+            _isFinalBlock = isFinalBlock;
+
+            // Note: We do not retain _bytesConsumed or _sequencePosition as they reset with the new input data
+            _stackFreeContainer = state._stackFreeContainer;
+            _lineNumber = state._lineNumber;
+            _lineBytePosition = state._lineBytePosition;
+            _currentDepth = state._currentDepth;
+            _maxDepth = state._maxDepth == 0 ? JsonReaderState.StackFreeMaxDepth : state._maxDepth;
+            _inObject = state._inObject;
+            _isNotPrimitive = state._isNotPrimitive;
+            _tokenType = state._tokenType;
+            _readerOptions = state._readerOptions;
+            _stack = state._stack;
+
+            _consumed = 0;
+            _totalConsumed = 0;
+            _isLastSegment = _isFinalBlock;
+            _isSingleSegment = true;
+            _currentPosition = default;
+            //_nextPosition = default;
+            _sequence = default;
+
+            IsValueMultiSegment = false;
+            ValueSpan = ReadOnlySpan<byte>.Empty;
+            ValueSequence = ReadOnlySequence<byte>.Empty;
+        }
+
+        /// <summary>
+        /// Read the next JSON token from input source.
+        /// </summary>
+        /// <returns>True if the token was read successfully, else false.</returns>
+        /// <exception cref="JsonReaderException">
+        /// Thrown when an invalid JSON token is encountered according to the JSON RFC
+        /// or if the current depth exceeds the recursive limit set by the max depth.
+        /// </exception>
+        public bool Read() => ReadSingleSegment();
+
+        private void StartObject()
+        {
+            _currentDepth++;
+            if (_currentDepth > _maxDepth)
+                ThrowHelper.ThrowJsonReaderException(ref this, ExceptionResource.ObjectDepthTooLarge);
+
+            _consumed++;
+            _lineBytePosition++;
+
+            if (_currentDepth <= JsonReaderState.StackFreeMaxDepth)
+                _stackFreeContainer = (_stackFreeContainer << 1) | 1;
+            else
+                _stack.Push(JsonTokenType.StartObject);
```

Comment:
I know that our style guidelines say that unbraced ifs are allowed here (since both sides have only one line and they're contained in the method root scope), but my personal rule is that they're only allowed in preambles when the body is a throw.

I merely mention that, in hopes that you also convert to "yeah, the braces do add a lot, but they are pretty noisy in precondition checks... maybe that's the right balance".